libs/api/libbpf_map.cpp

```diff
@@ -408,3 +473,120 @@ libbpf_bpf_map_type_str(enum bpf_map_type t)
 
     return _ebpf_map_display_names[t];
 }
+
+typedef struct perf_buffer
+{
+    std::vector<ebpf_map_subscription_t*> subscriptions;
+
+    perf_buffer() : ctx(nullptr), sample_cb(nullptr), active_subscription_count(0), free_inprogress(false) {}
+    ~perf_buffer();
+
+    void* ctx;
+    perf_buffer_sample_fn sample_cb;
+    size_t active_subscription_count;
+    bool free_inprogress;
+    std::condition_variable cv;
+    std::mutex lock;
+} perf_buffer_t;
+
+perf_buffer ::~perf_buffer()
+{
+
+    active_subscription_count = subscriptions.size();
+    free_inprogress = true;
+
+    for (auto& sub : subscriptions) {
+        ebpf_map_unsubscribe(sub);
+    }
+
+    subscriptions.clear();
+
+    {
+        std::unique_lock<std::mutex> lock_wait(lock);
+        // Wait for for all the notifications to be complete before freeing memory.
+        cv.wait(lock_wait, [this] { return this->active_subscription_count == 0; });
+    }
+}
+
+int
+perf_buffer_sample_fn_wrapper(void* ctx, int cpu, void* data, uint32_t size)
+{
+    UNREFERENCED_PARAMETER(cpu);
+    perf_buffer_t* perf_buffer = reinterpret_cast<perf_buffer_t*>(ctx);
+
+    if (perf_buffer->free_inprogress) {
+        std::lock_guard<std::mutex> lock(perf_buffer->lock);
+
+        if ((data == nullptr) || (size == 0)) {
+            --perf_buffer->active_subscription_count;
+        }
+    }
+
+    perf_buffer->sample_cb(perf_buffer->ctx, cpu, data, size);
+
+    if ((perf_buffer->free_inprogress) && (perf_buffer->active_subscription_count == 0)) {
+        perf_buffer->cv.notify_all();
+    }
+
+    return 0;
+}
+
+struct perf_buffer*
+perf_buffer__new(
+    int map_fd,
+    size_t /* page_cnt*/,
+    perf_buffer_sample_fn sample_cb,
+    perf_buffer_lost_fn lost_cb,
+    void* ctx,
+    const struct perf_buffer_opts* /* opts*/)
+{
+    ebpf_result result = EBPF_SUCCESS;
+    perf_buffer_t* local_perf_buffer = nullptr;
+
+    if ((sample_cb == nullptr) || (lost_cb == nullptr)) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Exit;
+    }
+
+    try {
+        std::unique_ptr<perf_buffer_t> perf_buffer = std::make_unique<perf_buffer_t>();
+        uint32_t ring_count = libbpf_num_possible_cpus();
+        ebpf_map_subscription_t* subscription = nullptr;
+        perf_buffer->ctx = ctx;
+        perf_buffer->sample_cb = sample_cb;
+
+        for (uint32_t cpu_id = 0; cpu_id < ring_count; cpu_id++) {
+            result = ebpf_map_subscribe(
```

Comment:
I do not think that this is necessarily the right approach. Why should we have one subscription per CPU? A single subscription for a perf_event_array will issue N async queries per CPU, whereas that for ring_buffer will issue a single async query.